### DISCUSSION TITLE
Nix-style "inherit"

### DISCUSSION THREAD

**Discussion by cueckoo** [cue-lang/cue#847] [Ideas]
_Originally opened by @nyarly in https://github.com/cuelang/cue/discussions/847_

I'm finding (and perhaps I'm discovering an anti-pattern) that I've got a lot of structs, often in a "function" style

```cue
foo: {
  parm: string
  result: #Def & { "parm": parm }
}

//used as
(foo & {parm: "bar}).result
```

One thing of note is the pattern of using a explicit string as a struct field name, to prevent CUE seeing it as a identifier and recursively trying to resolve it.

But when there's even a few parameters, it feels stuttery (a la Go style.)

Three other languages that occur to me have recognized this stylistic issue, and addressed it. ECMAScript simply allows bare names in `{ }` to be the equivalent of using the enclosing variable value and the object value. Rust does likewise.

Nix, on the other hand, has an `inherit` keyword, that takes a list of identifiers. The advantage of this feature is that inherit can also take a parenthesized identifier as a source of inheriting values, which means that a set (~= a struct) can be built up from other sets very easily.

For the above example, an ES/Rust style would look like
```cue
foo: {
 parm: string
 result: #Def & { parm }
 }
```

Nix style:
```cue
foo: {
  parm: string
  result: #Def & { inherit parm }
}
```

**Comment by cueckoo** [comment]
_Original reply by @verdverm in https://github.com/cuelang/cue/discussions/847#discussioncomment-504248_

Related to #484 and https://github.com/cuelang/cue/issues/69#issuecomment-739228231

- The bare name already has a semantic meaning, to embed the value into the struct
- The proposed "function" syntax may help, but depends on how you define it. Look into the `#0, #1` idea @mpvl brought up
- The Query Proposal is also relevant https://github.com/cuelang/cue/issues/165

`inherit` (or an embedding) cloud be achieved with a def, something like

```
#input: {
  p1: string
  p2: int
}

foo: {
  inputs: #input
  result: #Def & { inputs }
}
```

**Reply by cueckoo** [reply]
_Original reply by @nyarly in https://github.com/cuelang/cue/discussions/847#discussioncomment-504248_

Here's what I see, though: https://cuelang.org/play/?id=FfoY3gU-1rs#cue@export@cue

IOW, the embedded `inputs` doesn't see to carry the individual parameters into place.

**Reply by cueckoo** [reply]
_Original reply by @verdverm in https://github.com/cuelang/cue/discussions/847#discussioncomment-504248_

you need an extra `inputs: { <...> }` to make it work as expected

https://cuelang.org/play/?id=V9RZAqZLMZ4#cue@export@cue

**Comment by cueckoo** [comment]
_Original reply by @eonpatapon in https://github.com/cuelang/cue/discussions/847#discussioncomment-512121_

For those that don't know what does `inherit` in nix:

```nix
rec {
   foo = 1;
   bar = 2;
   b = { inherit foo bar; };
}
```

Is just a shorthand for:

```nix
rec {
  foo = 1;
  bar = 2;
  b = {
    foo = foo;
    bar = bar;
  };
}

 nix-instantiate --strict --eval test.nix
{ b = { bar = 2; foo = 1; }; bar = 2; foo = 1; }
```

As scoping rules are not the same than `cue`, `foo = foo` will not end up in a cycle error. It could if the `b` attributeset is defined as recursive (which is a bit like `cue` standard scoping rules):

```
rec {
   foo = 1;
   bar = 2;
   b = rec { foo = foo; bar = bar; };
}

 nix-instantiate --strict --eval test.nix
error: infinite recursion encountered, at /home/eon/test.nix:4:20
(use '--show-trace' to show detailed location information)
```

Using inherit in this case works as there is no self referencing attribute:

```
rec {
   foo = 1;
   bar = 2;
   b = rec { inherit foo bar; };
}

 nix-instantiate --strict  --eval test.nix
{ b = { bar = 2; foo = 1; }; bar = 2; foo = 1; }
```

Note that the keyword `inherit` is quite confusing for new comers.

In the context of `cue` I have found myself wanting to have a similar functionality when you want to copy a set of fields in a downward struct. Not only in `cue` you need to repeat the attributes/values you want to duplicate but you need also to avoid cycle errors with aliases or other methods:

```cue
{
	Foo=foo: 1
	Bar=bar: 2
	b: {
		foo: Foo
		bar: Bar
	}
}
```

**Reply by cueckoo** [reply]
_Original reply by @myitcv in https://github.com/cuelang/cue/discussions/847#discussioncomment-512121_

Could your example:

```
{
	Foo=foo: 1
	Bar=bar: 2
	b: {
		foo: Foo
		bar: Bar
	}
}
```

not be rewritten as:

```
{
	let x = {
		foo: 1
		bar: 2
	}
	x
	b: {
		x
	}
}
```

i.e. use `let` and embedding.

**Reply by cueckoo** [reply]
_Original reply by @eonpatapon in https://github.com/cuelang/cue/discussions/847#discussioncomment-512121_

Sure that would work also and is more concise. Maybe a bit more hard to read.

I think with `inherit` in nix, once you know exactly what it does, it is concise and quite readable. But I am probably biased on that since I'm used to it :)

Anyway I don't know if it would bring a lot of value to have something similar in `cue`. As you point out there is alternatives for this kind of situations (embedding is not possible in `nix`).

**Reply by cueckoo** [reply]
_Original reply by @nyarly in https://github.com/cuelang/cue/discussions/847#discussioncomment-512121_

The motivating case is something like:
```
z: {
	let x = {
		foo: int
		bar: int
	}
	x
	b: {
		x
	}
}

w: (z & {foo: 1, bar: 2}).b
```

If w evaluated to `{foo: 1, bar: 2}` I'd be happy with this solution, but the `let` doesn't work as an "anti-embed" here. (Which coinage only comes to mind because Nix calls its string interpolation "anti-quotation").

**Comment by mpvl** [comment]
One idea that recently came up is to allow:

```
#input: {
  p1: string
  p2: int
}

foo: {
  inputs: #input
  result: #Def & { : inputs }
}
```

where `:` without a field name means that the name of the last element in the reference is taken.

**Reply by ofek** [reply]
Is this functionality planned or is it still in the design phase?
